fixes #3917 - replace protected_attrs with strong parameters

domcleal opened the pull request:
Filtering of attributes has moved from the protected_attributes gem to
strong parameters in controller concerns, to be in line with current
Rails recommendations.

Concerns are shared between UI and both API controllers and list the
attributes using Foreman::ParameterFilter, which provides additional
features:
1. Registration of additional attributes from plugins, through the
   plugin API or `attr_accessible` for short term compatibility.
2. Re-use of permitted attribute lists for nested models (e.g. host
   and interface).
3. Combining of lists of attributes from all sources in a single
   permit call.
4. A small DSL for changing accepted parameters based on controller,
   action and UI/API type.

Plugins should either temporarily depend on protected_attributes to
continue to protect their models or call permit/use ParameterFilter
similarly: http://projects.theforeman.org/projects/foreman/wiki/Strong_parameters

Some UI changes were required to make nested model hash keys all
integers instead of "new_123456" etc, else strong parameters would
filter the entries out.

The Role model's builtin default has been moved from initialize to the
database as the removal of protected_attrs changed the initialisation
order.

domcleal committed:
fixes #3917 - replace protected_attrs with strong parameters

Filtering of attributes has moved from the protected_attributes gem to
strong parameters in controller concerns, to be in line with current
Rails recommendations.

Concerns are shared between UI and both API controllers and list the
attributes using Foreman::ParameterFilter, which provides additional
features:

1. Registration of additional attributes from plugins, through the
plugin API or `attr_accessible` for short term compatibility.
2. Re-use of permitted attribute lists for nested models (e.g. host
and interface).
3. Combining of lists of attributes from all sources in a single
permit call.
4. A small DSL for changing accepted parameters based on controller,
action and UI/API type.

Plugins should either temporarily depend on protected_attributes to
continue to protect their models or call permit/use ParameterFilter
similarly: http://projects.theforeman.org/projects/foreman/wiki/Strong_parameters

Some UI changes were required to make nested model hash keys all
integers instead of "new_123456" etc, else strong parameters would
filter the entries out.

The Role model's builtin default has been moved from initialize to the
database as the removal of protected_attrs changed the initialisation
order.

lzap commented:
Please do not merge this immediately after it's reviewed, announce a day for the merge at least three working days in advance so plugin authors can prepare their own PRs. I'll be doing this for discovery this week. (Edit: I am assuming this is not compatible and will break most plugins.)

Looks like we need to turn off codeclimate's "Similar Code" thing, it's annoying.

domcleal commented:
> (Edit: I am assuming this is not compatible and will break most plugins.)

You can probably retain compatibility in most cases by depending on protected_attributes if your plugin uses it, which you can/should do today - not after this is merged.  More details are on the linked wiki page.  If you have specific requests for extra compatibility, please comment with the details and I'll see what I can improve.

lzap commented:
We have a generic searching API, is it worth creating a generic param filter for these (`index`) actions? Something like:

``` ruby
  def search_params
    params.permit(:search, :order, :page)
  end
```

domcleal commented:
@lzap yes, can do - though where is it used? The parameters are only filtered on updates/creates normally because they're used for mass-assignment.  Search parameters on indexes shouldn't be mass-assigned - continue to use `params[:search]` etc.

lzap commented:
I see, scratch it.

Great work, I finished refactoring of discovery, was almost a day of work, but thanks to your wiki I was able to start quickly. I roughly tested PXE provisioning and it works just fine.

My only concern is documentation for the `:top_level_hash` option, if you could elaborate it. I had to use it in discovery, didn't work without it (feel free to review my plugin-side changes - https://github.com/theforeman/foreman_discovery/pull/285).

Amount of changes is huge, I can barely test everything, but since core tests are green I can read through the patch and just merge it if there are no objections. Since this is security-facing I'd rather see more than one reviewers here.

domcleal commented:
> My only concern is documentation for the :top_level_hash option, if you could elaborate it. I had to use it in discovery, didn't work without it

Yeah, it's necessary in a few places where the controller name differs from the model name, particularly with LookupKey + its STI derivatives, which have three different controller names between the UI and API versions.  I'll see if it can be set automatically from the controller and add documentation.

The functional tests do indeed test almost everything at the basic level. The likeliest issues are in nested attributes, e.g. parameters or lookup values within hosts, domains etc, because these required the largest changes from attr_accessible which didn't work the same way for nested resources.  The users controllers also have a substantial refactoring of params whitelisting for non-admin users - though there are new tests added to cover this.

lzap commented:
Ok thanks for explanation, I will go ahead and go through the code and then merge.

dLobatog commented:
[test] - to check if Katello is fine or not

lzap commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -53,7 +53,7 @@ function fix_template_context(content, context) {
   var regexp  = new RegExp('new_' + assoc, 'g');
-  return content.replace(regexp, "new_" + new_id);
+  return content.replace(regexp, new_id);
 }
```
Still long read for me, but is this related?

domcleal replied:
Yes.  When strong_attribute filters nested parameters (in the form `:host => {:host_parameters_attributes => {'1' => {:name => 'foo', :value => 'bar'}}`), then you call `params.require(:host).permit(:host_parameters_attributes => [[:name, :value]])` to indicate that it's an array of hashes with the keys name, value.

Even though the input is a hash (as generated by the Rails nested forms), strong parameters understands it's essentially an array, but _only_ if all of the keys are numeric.  If there are any non-numeric keys then the entire sub-hash is filtered out.

The change here ensures that when a new child object is created in our forms (e.g. a new NIC or parameter), that it has a solely numeric key and not "new_123456789".  The changes to serializeForm() are similar and filter out the child form templates - if the form templates are included then we end up with numeric keys for all of the existing and new nested items, but also a "new_interface" item that causes the whole hash to be filtered.

lzap commented on app/controllers/concerns/api/v2/taxonomies_controller.rb:
```diff
@@ -124,4 +124,8 @@ def find_taxonomy
+
+  def resource_params
+    public_send("#{taxonomy_single}_params".to_sym)
+  end
```
What is the reason for the change?

domcleal replied:
This concern is used in both OrganizationsController and LocationsController classes, one with `organization_params` and the other with `location_params`, so this just calls the right one depending on which controller type we're in.

lzap commented on app/controllers/concerns/foreman/controller/parameters/external_usergroup.rb:
```diff
@@ -0,0 +1,17 @@
+      Foreman::ParameterFilter.new(::ExternalUsergroup).tap do |filter|
+        filter.permit :usergroup, :usergroup_id, :usergroup_name
+        filter.permit_by_context :name, :auth_source_id, :auth_source_name, :nested => true
+        filter.permit_by_context :id, :_destroy, :ui => false, :api => false, :nested => true
```
Only a subjective opinion, for better git merging, I'd prefer one item per line. It's a chance to set the standards. I tend to prefer practical approach to appearance. Just for comparison, this one would look like:

``` ruby
filter.permit
  :usergroup,
  :usergroup_id,
  :usergroup_name
filter.permit_by_context
  :name,
  :auth_source_id,
  :auth_source_name,
  :id,
  :_destroy,
  :ui => false,
  :api => false,
  :nested => true
```

I understand I am part of minority in this case. :-)

lzap commented on app/controllers/concerns/foreman/controller/users_mixin.rb:
```diff
@@ -12,40 +11,13 @@ def resource_scope(options = {})
+      # Remove locale from the session when set to "Browser Locale" and editing self
+      session.delete(:locale) if params[:user][:locale].try(:empty?)
     end
   end
```
Note for myself: Check if this is covered with tests.

lzap commented:
What a monster patch, I don't have any important comments, good work.

dLobatog commented:
Tests are  currently failing because foreman-docker is a Katello depedendency which is not converted to strong_params yet.

lzap commented:
Thanks for explanations, I set RTM unless there are objections.

domcleal commented:
The changes to top_level_hash haven't yet been made, which is why it was on WoC. I also haven't addressed your comment about listing attributes one a line.  Do you want these left as-is?

lzap commented:
> The changes to top_level_hash haven't yet been made, which is why it was on WoC. I also haven't addressed your comment about listing attributes one a line.  Do you want these left as-is?

I haven't recognized top_level_hash needs a change, the syntax is
acceptable and it was rather observation.

The one line vs multiline is very hard one. Honestly, I haven't expected
anyone to like it. But if there are no objections, I see this very
practical and I am all for it!

Decide if the two changes are worth the effort yourself and set the
flags appropriately, sorry for the noise then.

## 

Later,
 Lukas #lzap Zapletal

domcleal commented:
I've updated the PR to:
1. Mostly remove top_level_hash, which is now derived from the controller name so it better matches the standard hash. The options hash that contained :top_level_hash has been removed and it's now an option on the `filter_params` method as it still has a use in a few controllers (trends, dashboards, puppetclasses) with non-standard params layouts.
2. Reorder attributes in alphabetical order and mostly one per line.  Variants of the same association are kept on the same line (e.g. operatingsystem_id, operatingsystem_name).

lzap commented:
Merged as 12612809c5deb885b13ceaf36b6a147c21688eb5, thank you!

lzap commented:
It's in! Some PRs might need a rebase and plugins should be fixed now. Nice one.
